### DISCUSSION TITLE
! has shocking behavior

### DISCUSSION THREAD

**Discussion by kjpgit** [dotnet/csharplang#5968] [Q&A] [answered]
**EDIT: Found this awesome project https://github.com/tom-englert/Nullable.Extended
It warns on usage of `!` unless you have a special comment to suppress the warning.  Check it out!**  (And hopefully it becomes a standard warning you can request from the compiler!)

I wrote this line of code today in .net 6

```
var db = new CPStorage(Environment.GetEnvironmentVariable("CPS_TABLE_NAME")!);
```

And then started up my app to verify the error that would be thrown since this env var doesn't exist yet, and CPStorage() requires a non-null string.  (And I was as expected getting a nullability warning before I added the !) And to my dismay, it started up with no error.  I was shocked.

I have used swift, where ! will trap at runtime if it is null.  I expected c# not to trap (Exit), but throw a null exception.

If I wanted to turn off a compiler warning and do something unsafe, I would have used a `pragma` or `unsafe` keyword.  Not this.

I have been slogging through asp.net for a new project, and have been willing to overlook a lot of struggles because I think C# is a decent language, but this has me worried I'm making a bad bet.

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-infrastructure-libraries
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

I wrote a line of code today,

```

<table>
  <tr>
    <th align="left">Author:</th>
    <td>kjpgit</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-Infrastructure-libraries`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by kjpgit** [comment]
I seem to have hit ctrl-enter or shfit-enter very early when typing my code (annoying feature to be honest), so it prematurely called the bots.

**Comment by bartonjs** [comment]
`!` in C# is just the Jedi mind trick for "there's no null here...".  It doesn't cause a throw to happen, just makes a "this might be null..." warning go away.

If it's used in the middle of a member invocation expression (e.g. `Environment.GetEnvironmentVariable("CPS_TABLE_NAME")!.Substring(0)`) then the member invocation is what will throw a `NullReferenceException`, not the `!`.

For the line of code as written I'd expect the `CPStorage` constructor to throw an `ArgumentNullException`... unless the `CPStorage` constructor wasn't marked as non-null-accepting, in which case the suffix `!` has no effect in code.

In C# the suffix `!` is called the `null-forgiving operator`: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving

> Available in C# 8.0 and later, the unary postfix ! operator is the null-forgiving, or null-suppression, operator. ... The null-forgiving operator has no effect at run time. It only affects the compiler's static flow analysis ...

**Comment by danmoseley** [comment]
@kjpgit what tools are you using? If you use Visual Studio I suggest you hit debug to see what's happening.

**Comment by kjpgit** [comment]
So let's see how to write safe, modern c# code in 2022:

```
class Storage
{
    public Storage(string required_table_name) { }
}

// Not this way
new Storage(Environment.GetEnvironmentVariable("CPS_TABLE_NAME")!);

// Not this way
new Storage((string)Environment.GetEnvironmentVariable("CPS_TABLE_NAME"));

// Yep, this is clearly the idiomatic and simple way
using static correctness_is_always_job_one;
class correctness_is_always_job_one
{
    public static T __unwrap<T>(T? obj) {
        if (obj is null) {
            throw new ArgumentNullException();
        }
        return obj;
    }
}
// Errors do not pass silently
// Do not violate principle of least surprise
// Correctness is job 1 of business software
new Storage(__unwrap(Environment.GetEnvironmentVariable("CPS_TABLE_NAME")));
```

I vote for a `#pragma bang_enforces` directive (or compiler option preferably) where `!` does not violate the preconditions of the callee.  If I wanted to YOLO I'd use python.

The SHORTEST code had better be the correct thing, because that's what everyone will do.  I'm guessing people are just using ! everywhere because the compiler isn't don't good enough control flow analysis, but there should be a different, longer symbol (or system function) for that and it should be marked as `unsafe` in modern codebases.

 I can not express how disappointed I am with this "new philosophy" of the c# stewards, where unsafe code is now the easiest and shortest path.  Apple's Swift language got this right with the ! unwrap operator (although I'd prefer an exception instead of terminating the thread, for server software obviously).
  
  Please transfer this issue to the language repo if appropriate.  Thank you.

**Reply by HaloFour** [reply]
See: #2145

Initially the team is considering it only for parameters but there is a discussion about maybe expanding it to expressions.

**Reply by kjpgit** [reply]
Yes, I was aware of that before I posted this issue, and I find it deeply troubling.  The hole keeps getting dug deeper, instead of fixing the source problem (which is an easy fix).
I suggest the team spend a week and write a short Swift program (seriously).  You will then see how much safer it is to just make `obj!` enforce, and `__assume_not_null(obj)` not enforce.

**Reply by HaloFour** [reply]
> I suggest the team spend a week and write a short Swift program (seriously).

The team are well versed in Swift and dozens of other languages.  They are very aware that NRTs are a compromise.  That is what happens when you try to introduce a massive change to the type system 18 years after a language ships.  Swift had the advantage of having non-null reference types out the gate and could make very different choices.

> You will then see how much safer it is to just make `obj!` enforce

That will not happen.  The feature has already shipped as a part of NRTs which were designed very explicitly to not have any runtime semantics.  That operator exists only to inform flow analysis that you, as the developer, insist that the value cannot be `null` at that point.

**Reply by Richiban** [reply]
@kjpgit 

>  how much safer it is to just make `obj!` enforce

I think your shock comes from a misunderstanding of why the dammit operator was introduced in the first place. It's not an "assert this thing is not null" operator it's an "I know what I'm doing, turn off null tracking for this value" operator.

It's necessary in situations like the following where, unlike a language like Kotlin, C# has no `lateinit` operator:

```cs
class C
{
    private string _data = null!; // If the ! operator enforced at runtime then 
                                  // this pattern would become impossible

    public void Init()
    {
        _data = "some string";
    }
```

**Reply by kjpgit** [reply]
> "I know what I'm doing, turn off null tracking for this value"

Which is 100% the opposite of the original c# design philosophy, to be be safer than c++.  That is where my shock is coming from.
Why is there garbage collection by default?  Why are variables always initialized by default?  Because the "I know what I'm doing" attitude fails quickly for business software, and those lack of guard rails were proven to be too dangerous.   The fact that you can do this with a single character, which is hard to grep for, in modern c# makes me feel like it is not for me.  I don't even know who the target audience of c# is at this point.  It seems to be (re)starting the same course as c++ did.  

C# also cared about programmer portability, making it at least similar to c++.  You are doing something 100% opposite of Swift here, with no valid reason that I can see to overload the ! instead of making a different symbol or function name.

edit: also, I don't fully understand your example, sorry.  It seems like _data would just be declared as optional, and you would provide accessors or properties that can cast it to a non nullable.  And you could choose to have those accessors do a runtime check (!), or a __trust_me(obj) check with no overhead.

**Reply by CyrusNajmabadi** [reply]
> with no valid reason that I can see 

We've stated the reasons many times in the past. You may not agree with them, but they are there. 

NRT faced an enormous uphill battle to even get people to consider it. We had to do a lot of work to make it palatable in the slightest for the .net community. And that community is broad with billions of lines of code that are mission critical that they want to keep working and they do not want to deal with a lot of pain over. 

To that end, we needed a very easy and simple way for people to adopt nrt, but also tell them system: your analysis isn't advanced enough to prove what I'm doing is safe. I know more (for example things that may be true about a particular domain, or about how a set of components work together), so here's a short and simple easy for me to say: I know what I'm doing *disable* this analysis. 

This made the feature palatable for people to adopt in their codebases, and it was necessary to get the adoption and to prevent this feature from going nowhere.

**Reply by Richiban** [reply]
I guess I don't fully understand your objection here... is it that they chose `!` for the operator, which doesn't match what `!` does in Swift? Or is it that they haven't provided a short, easy and built-in way of asserting (at runtime) that a value is not null?

As others have mentioned in this thread, NRTs were an unbelievably complex addition to the language considering it'd been around for almost 20 years and, honestly, I think they've done the best possible job given the circumstances; any other (stricter) route would have involved forking not only the language but the runtime *and* the base class libraries.

When they added NRTs the need for a "dammit" operator _was discovered_ since, due to the fact that NRTs in C# are little more than compiler-integrated null tracking, in some situations the programmer really does know more than the compiler about the nullity of a particular variable.

As for the syntax conflict with Swift, well, they're under no obligation to follow the syntax choices of other languages, as those languages are under no obligation to follow C#.

**Reply by kjpgit** [reply]
> I guess I don't fully understand your objection here... is it that they chose `!` for the operator,

Yes, exactly, I don't know how many times I can repeat myself.  The current no-runtime-check-but-shutup-compiler-warning feature should be called `__assume_not_null(obj)` (or something else if you want to annotate "late init variables", which you might already be able to do with a pragma disable warnings) and the `!` should be a runtime check.  So the shortest path, that green-field projects will use which is !, will be safe.

> Or is it that they haven't provided a short, easy and built-in way of asserting (at runtime) that a value is not null?

Yes, I do find this incredibly silly that I had to write my own helper function to do this.  It really makes me question the future of the language that this was "missed" TBH.  It seems like green field projects weren't even considered with NRT, almost as if you guys expect to only maintain existing codebases and not add new developer mindshare...

**Reply by HaloFour** [reply]
The `!` operator has shipped, its meaning is not going to change, nor will dialects be added via pragma to change it.  At this point the discussion would be around what new features can be added to the language to improve null safety.  Relitigating past decisions will not go anywhere.

**Reply by CyrusNajmabadi** [reply]
> Yes, exactly, I don't know how many times I can repeat myself. The current no-runtime-check-but-shutup-compiler-warning feature should be called __assume_not_null(obj)

That would be unpalatable for our existing ecosystem and community.  We do not want to hobble features this important with baggage that will prevent them from getting adopted.  That defeats teh purpose of NRT.

**Reply by kjpgit** [reply]
@HaloFour can we have a compiler warning if `!` operator is being used?  I would turn that on immediately for my green-field project and have the team fix those warnings by a) using our `__unwrap<T>` helper method, or b) putting warning-silence pragmas around the code blocks that truly need it (which there probably won't be in this project).

**Reply by CyrusNajmabadi** [reply]
>like green field projects weren't even considered with NRT, almost as if you guys expect to only maintain existing codebases and not add new developer mindshare...

There is literally 8-9 *orders of magnitude* more existing code than green field code.  We are focusing on teh areas that need the most help.  It's not htat we didn't consider it, it's just that it's vastly less important than the enormity of existing code out there that needs these features and comes with incredibly constraints on what is acceptable or not.

**Reply by sharwell** [reply]
> can we have a compiler warning if ! operator is being used? 

Yes, you can definitely create a custom analyzer and install it in your solution to block the use of `!` (and even provide a code fix to convert such use to the helper/extension method you use).
https://docs.microsoft.com/en-us/visualstudio/extensibility/getting-started-with-roslyn-analyzers?view=vs-2022

**Reply by CyrusNajmabadi** [reply]
> @HaloFour can we have a compiler warning if ! operator is being used?

Yes.  It would take just a few minutes to write an analyzer to do that for yourself.  Just use `context.RegisterSyntaxNodeAction(..., SyntaxKind.SuppressNullableWarningExpression);` and you can report a warning/error on every usage in your project.

**Reply by kjpgit** [reply]
For the record, the "global usings" with "static=True" feature did make it convenient to expose our  `__unwrap<T>` helper function to all source code files.  And I do commend the effort toward increasing null safety.  I will look at the roslyn links when I have time.  Lastly, in what repo should I file a request to add an official compiler warning?

**Reply by CyrusNajmabadi** [reply]
> Lastly, in what repo should I file a request to add an official compiler warning?

github.com/dotnet/roslyn-analyzers would be the repo.  However, it is highly unlikely we would add a feature to warn on a language feature being used for its intended purpose.

**Reply by kjpgit** [reply]
Thanks.  Filed https://github.com/dotnet/roslyn-analyzers/issues/5934

**Reply by RikkiGibson** [reply]
The closest analogue to C#'s postfix `!` that I'm aware of is probably the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) in TypeScript.

**Reply by kjpgit** [reply]
FYI This project with [7000 nuget downloads ](https://github.com/tom-englert/Nullable.Extended) does what I want... warns about all uses of ! unless you suppress it with a special comment.  So we heavily discourage obj! (which removes safety, clearly some language features are not as safe as others) and use our `__unwrap(obj)` instead.    Yay, I have a working solution and feel the Zen now. (Well, other than relying on a random stranger's nuget package...WCGW?)

I don't have a strong opinion on the proposed !! thing.   For marking / declaring arguments,  I can see how it would be really short for existing code that has to support legacy callers.  For use in code/expressions, meh, I like our our __unwrap looks -- easy to grep, just the right amount of friction that makes you think "should I do an explicit check here?", but at least "!!" might be able to be grepped for... except it would overlap with uses of !! in arguments... so meh, I don't care either way about the !!.

**Comment by danmoseley** [comment]
@kjpgit nobody here is saying you should write that code. We don't yet understand what the problem is.

**Comment by danmoseley** [comment]
Starting at the start again.
`var db = new CPStorage(Environment.GetEnvironmentVariable("CPS_TABLE_NAME"));`

If this environment variable does not exist, GetEnvironmentVariable will return null as you know. What happens now depends on the implementation of CPStorage:
1. If it accepts null then execution continues normally
1. If it does not, but it was built with a [nullable annotation context](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts) (which essentially means it's API has been annotated) then you will get a compile time warning, because GetEnvironmentVariable has been annotated to potentially return null. You can either suppress this warning with `!` or change your code so that it can't pass in null.
1. If it does not, but it's not participating in nullable annotations, then you will get no warning and only a run time error.

Does that match your expectations? In no circumstances do you need to write the extended code you have pasted.

**Comment by huoyaoyuan** [comment, marked as answer]
You can use `?? throw` expression, or you can make the `__unwrap` extension method.

`!` is designed for when nullability is obvious to human, but not provable for the compiler. It's intentionally not checking for performance reason, for example the usage in `ImmutableArray<T>`.

The nullable reference type feature of C# 8 is designed to be pure compile time check, to avoid breaking existing code as possible.

**Reply by kjpgit** [reply]
Just to be clear,  I assume you are suggesting `obj ?? throw new ArgumentNullException()` ?  Given the choice between that and `obj!`, which do you think the average programmer is going to do?  Which do you think programmers who care about trapping errors early are going to feel happy about?

My suggestion is simple: 
* Make `obj!` actually enforce at runtime
* Make a `T System.__assume_not_null<T>(T? t)` for the "elite 1% performance junkies" who want to disable a warning, e.g. `__assume_not_null(obj)`

We have a language for performance junkies, it's C++.  I've lived in that world, where the defaults are unsafe.  It's terrible.

**Reply by CyrusNajmabadi** [reply]
> Make obj! actually enforce at runtime

This would defeat the purpose of nrt and make it's literal non-starter for a vast section of our user base. Nrt was accepted because the claim can honestly and accurately be made that it has no runtime impact, and this can be adopted into existing codebases without risk. 

> We have a language for performance junkies, it's C++

You are fundamentally misunderstanding. This is not about performance. This is about *risk* management for teams contemplating using nrt. 

If they adopt nrt and now get runtime exceptions for code that worked before just by using nrt, then that is a complete non-starter.

**Reply by sharwell** [reply]
This is the correct answer. The language provides `!` as a suppression for cases where nullable reference types doesn't support the semantics of the program and the value is known to not be null, and provides `?? throw ...` for cases where a truly nullable value is being converted for use in a non-nullable context. The option used an any given location depends on the program semantics.

**Comment by B1Z0N** [comment]
Also, there's supposedly will be `!!` operator in `dotnet 7.0`/`c# 11` that solves just what you were talking about here. See [this](https://github.com/dotnet/csharplang/discussions/5735#discussioncomment-2152891) for more info.

**Comment by TahirAhmadov** [comment]
Why not just write an extension method?
```cs
public static class _Object_Extensions
{
  [MethodImpl(AggressiveInlining)]
  public static T NotNull<T>(this T? This) where T : class => This ?? throw new NullReferenceException();
}

string? s = null;
Foo(s.NotNull());
```

**Reply by kjpgit** [reply]
I didn't realize it would work if `this` is null, based on knowledge of existing APIS like String.IsNullOrEmpty().  Will consider it, as it would be practically equivalent to Rust's `Option.unwrap()`.  But would it potentially conflict in the future if microsoft adds its own method with the same name?   I'm a bit shocked I was able to use NotNull from another source file without adding a `using`.

**Reply by CyrusNajmabadi** [reply]
> But would it potentially conflict in the future if microsoft adds its own method with the same name?

This is likely never going to happen.  And this would make ever using an extension method problematic, and that issue exists today.

> I'm a bit shocked I was able to use NotNull from another source file without adding a using.

Is the containing static class in scope in the location where you're using the extension method?

**Reply by kjpgit** [reply]
My bad, yeah they were both in the default namespace. After I put the NotNull class in a different namespace, I got an error, and had to add a using, which is good.

I guess I'm not 100% clear what you're saying about "issue that exists today", but just for the sake of argument -- if microsoft did add their own extension method, or direct method (edit: probably not possible for something where this is null?), with the same name, would I be able to disambiguate and still prefer my own extension method that predated theirs?

**Reply by jmarolf** [reply]
I am assuming that if we allowed you to put `!!` and it meant "Throw null reference exception" You would be in favor?

Essentially this:

```csharp
var db = new CPStorage(Environment.GetEnvironmentVariable("CPS_TABLE_NAME")!!);
```

Would be shorthand for this:

```csharp
var db = new CPStorage(Environment.GetEnvironmentVariable("CPS_TABLE_NAME") ?? throw new NullReferenceException());
```

If so, good news! this is something the language design team is thinking about doing.

**Reply by kjpgit** [reply]
@jmarolf  I don't have a strong opinion on the proposed !! thing (as long as I can warn/discourage ! in my codebase). For marking / declaring arguments, I can see how it would be really short for existing code that has to support legacy callers. For use in code/expressions, meh, I like how our __unwrap looks -- super easy to grep, just the right amount of friction that makes you think "should I do an explicit check here?"  "Am I ok with this small runtime performance hit?"   The symbol "!!" might be able to be grepped for, unlike '!', ... except it would overlap with uses of !! to annotate arguments... so meh, I don't care much either way about the !!.

It seems like bikeshedding, what syntax you want, you'd have to make a lot of real example code to see if you like the look of !! yelling !! at !! you !! vs `unwrap()` or `NotNull` -- which as this thread is showing, users have the option on what they want to call it... which may or not be a good thing, but I'm cautiously optimistic  :-)

**Reply by TahirAhmadov** [reply]
> I guess I'm not 100% clear what you're saying about "issue that exists today", but just for the sake of argument -- if microsoft did add their own extension method, or direct method (edit: probably not possible for something where this is null?), with the same name, would I be able to disambiguate and still prefer my own extension method that predated theirs?

Please refer to the docs on method resolution, or just experiment - create a sample assembly with this method, then create another project referencing the first assembly and also defining this method, and see what happens.
